homeassistant/components/device_tracker/__init__.py

```diff
@@ -460,6 +463,46 @@ def async_update(self):
             self._state = STATE_HOME
             self.last_update_home = True
 
+        # lookup vendor string for mac address using api.macvendors.com
+        if not self.vendor:
+            self.vendor = yield from self.get_vendor_for_mac()
+
+    @asyncio.coroutine
+    def get_vendor_for_mac(self):
+        """Try to find the vendor string for a given MAC address."""
+        # can't continue without a mac
+        if not self.mac:
+            return None
+
+        # prevent lookup of invalid macs
+        if not len(self.mac.split(':')) == 6:
+            return 'unknown'
+
+        # we only need the first 3 bytes of the mac for a lookup
+        # this improves somewhat on privacy
+        oui_bytes = self.mac.split(':')[0:3]
+        # bytes like 00 get truncates to 0, API needs full bytes
+        oui = '{:02x}:{:02x}:{:02x}'.format(*[int(b, 16) for b in oui_bytes])
+        url = 'http://api.macvendors.com/' + oui
+        try:
+            with async_timeout.timeout(5, loop=self.hass.loop):
+                resp = yield from self.hass.websession.get(url)
+            # mac vendor found, response is the string
+            if resp.status == 200:
+                vendor_string = yield from resp.text()
+                return vendor_string
+            # api 404 == mac vendor not found
+            elif resp.status == 404:
+                # set something other then None to prevent retry
+                return 'unknown'
+            # request failed, set None to try another time
+            else:
```

Comment:
I refactored this bit because the logic didn't make sense. Feel free te review and comment again if needed.